src/Common/src/System/Runtime/InteropServices/NativeBuffer.cs

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace System.Runtime.InteropServices
+{
+    using System;
+    using System.Runtime.CompilerServices;
+    using Microsoft.Win32.SafeHandles;
+
+    /// <summary>
+    /// Wrapper for access to the native heap. Dispose to free the memory. Try to use with using statements.
+    /// Does not allocate zero size buffers, and will free the existing native buffer if capacity is dropped to zero.
+    /// 
+    /// NativeBuffer utilizes a cache of heap buffers.
+    /// </summary>
+    /// <remarks>
+    /// Suggested use through P/Invoke: define DllImport arguments that take a byte buffer as SafeHandle or IntPtr.
+    /// NativeBuffer has an implicit conversion for both.
+    /// 
```

Comment:
Can we enforce SafeHandle always? I'd rather take the extra hit per P/Invoke than have to audit every use via IntPtr (which is really hard to see with an implicit conversion calling DangerousGetHandle, btw).